Add munit clues tutorial (#14)
scalacenter/docs.scala-lang: _overviews/toolkit/munit-clues.md

{% include markdown.html path="_markdown/install-munit.md" %}

> In MUnit tests outputs are presented to the user in different ways depending on used assertion.

Reviewer: ```suggestion
In MUnit tests outputs are presented to the user in different ways depending on which assertions are used.
```

For example `assertEquals` failure returns not only stack trace, 

> but also by default includes reason of the failure: `values are not the same` along with differences between two values.

Reviewer: ```suggestion
but also by default includes the reason of the failure: `values are not the same` along with differences between the two values.
```

> but also by default includes reason of the failure: `values are not the same` along with differences between two values.

Reviewer 2: The first sentences should be a summary of the tutorial.

Something along those lines:
```
In MUnit we can use `clues` to get better error reports. There are two kind of clues:
- the `clue` argument of the `assert` method to print an explanation of the error
- the `clue` method to choose which values to print in the error report
```

Then we can have two parts.

## Using the `clue` argument

## Using the `clue` method

Reviewer 3: Optionally, there could also be a short introduction to the problem, to get the user to feel that we are speaking exactly about the problem that he came to us with. Like
```
Normally, the MUnit will only provide the `values are not the same` error message. It's often not enough for making readable test reports and we can use `clues` to get better error reports. There are two kind of clues:
- the `clue` argument of the `assert` method to print an explanation of the error
- the `clue` method to choose which values to print in the error report
```

{% tabs munit-assert-equals %}

> {% tab 'Scala CLI' %}

Reviewer 2: But also, the tutorial should be self-contained. One user should not need to read more than one tutorial to solve its problem.

So I think it is actually better if the first code-block contains a full example. But then the subsequent examples can only contain the `assert`.

```scala
//> using toolkit

class MyTests extends munit.FunSuite:
  test("string comparison") {
    assertEquals("JohnSmith", "John Smith")
  }
```
{% endtab %}
{% tab 'sbt and Mill' %}
```scala
class MyTests extends munit.FunSuite:
  test("string comparison") {
    assertEquals("JohnSmith", "John Smith")
  }
```
{% endtab %}
{% endtabs %}

Which will result in:
```
# ==> X MyTests.string comparison munit.ComparisonFailException: 
#   ./example/MyTests.test.scala:7
# 4:  test("string comparison") {
# 5:    assertEquals("JohnSmith", "John Smith")
# 6:  }
# values are not the same
# => Obtained
# JohnSmith
# => Diff (- obtained, + expected)
# -JohnSmith
# +John Smith
```

Sometimes we may want to change it to better suit our needs.

> In MUnit there are two ways to achieve it. First one allows us to directly provide

Reviewer 4: ```suggestion
In MUnit, there are two ways to achieve it. The first one allows us to directly provide
```

> custom message to assertion functions as an additional parameter, for example:

Reviewer 4: ```suggestion
a custom message to assertion functions as an additional parameter, for example:
```

{% tabs munit-assert-clue-string %}
{% tab 'Scala CLI' %}
```scala
//> using toolkit

class MyTests extends munit.FunSuite:
  test("price after discount can't be lower than 0") {
    val price = 10.0
    val discount = 15.0
    assert(
      price - discount >= 0,
      s"price cannot have negative value: ${price - discount}"
    )

>   }

Reviewer 2: This should be the first code block.

```
{% endtab %}
{% tab 'sbt and Mill' %}
```scala
class MyTests extends munit.FunSuite:
  test("price after discount can't be lower than 0") {
    val price = 10.0
    val discount = 15.0
    assert(
      price - discount >= 0,
      s"price cannot have negative value: ${price - discount}"
    )
  }
```
{% endtab %}
{% endtabs %}

Now the test output contains a message when test fails:
```
# MyTests:
# ==> X MyTests.price after discount can't be lower than 0  munit.FailException: 
#   ./example/MyTests.test.scala:7 price cannot have negative value: -5.0
# 9:       s"price cannot have negative value: ${price - discount}"
# 10:    )
# 11:  }
```

The other way to change output messages is to use `clue` method for those arguments which values we want to be printed

> in case of failure:

Reviewer 4: This sentence is not very clear, could you maybe reword it or break it into multiple sentences? It could be useful to show an example without using `clue`, involving a complex expression, to show that the default report does not help to focus on the important parts of the expression, hence the need for `clue`.
